java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestFrameworkUtils.java

```diff
@@ -52,51 +52,93 @@ public static TestItem resolveTestItemForClass(IType type) throws JavaModelExcep
+            boolean checkHierarchy) {
+        if (!IAnnotatable.class.isInstance(member)) {
+            return Optional.empty();
+        }
 
-            final String name = annotation.getElementName();
+        final IJavaProject javaProject = member.getJavaProject();
+        if (javaProject == null) {
+            return Optional.empty();
+        }
 
-            String[][] fullNameArr = null;
-            if (IType.class.isInstance(member) && member.getDeclaringType() == null) {
-                fullNameArr = ((IType) member).resolveType(name);
-            } else {
-                fullNameArr = member.getDeclaringType().resolveType(name);
-            }
-            if (fullNameArr == null) {
-                final ICompilationUnit cu = member.getCompilationUnit();
-                if (cu != null && cu.getImport(memberAnnotation).exists()) {
-                    return Optional.of(annotation);
-                } else {
-                    return Optional.empty();
+        final IType declaringType = member.getDeclaringType();
+        if (declaringType == null) {
+            return Optional.empty();
+        }
+
+        final IAnnotatable annotatable = (IAnnotatable) member;
+        try {
+            for (final IAnnotation annotation : annotatable.getAnnotations()) {
+                final IType annotationType = getResolvedType(annotation.getElementName(), declaringType, javaProject);
+                if (annotationType != null) {
+                    if (annotationToSearch.equals(annotationType.getFullyQualifiedName())) {
+                        return Optional.of(annotation);
+                    }
+                }
+                if (checkHierarchy) {
+                    final Set<IType> hierarchy = new HashSet<>();
+                    if (matchesInAnnotationHierarchy(annotationToSearch, annotationType, javaProject, hierarchy)) {
+                        return Optional.of(annotation);
+                    }
                 }
             }
-            final String fullName = Arrays.stream(fullNameArr[0]).collect(Collectors.joining("."));
-            return fullName.equals(memberAnnotation) ?
-                Optional.of(annotation) : Optional.empty();
         } catch (final JavaModelException e) {
+            // Swallow the exception
             return Optional.empty();
         }
+        return Optional.empty();
     }
 
-    public static boolean hasAnnotation(IMember member, String annotation) {
-        return getAnnotation(member, annotation).isPresent();
+    /**
+     * Check the given {@link IMember} has annotated with the given annotation string.
+     *
+     * @param member the {@link IMember} to search.
+     * @param annotationToSearch The annotation string.
+     * @param checkHierarchy Specify whether to search the whole annotation hierarchy.
+     */
+    public static boolean hasAnnotation(IMember member, String annotationToSearch, boolean checkHierarchy) {
```

Comment:
Would `Inherited` be more proper than hierarchy? e.g. `checkInherited`

Reply:
`includeInherited` looks even better.

Reply:
Hmm, actually the annotation we treated here is not inherited. That's why it is called meta-annotation.

Reply:
exactly, not inherited. From the saying meta-annotation, it's annotation of an annotation, and it is, more like aggregation/combination.  
Anyway, use whatever proper name you like.